src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/RefAPISource.java

```diff
@@ -61,40 +73,66 @@ public static RefAPISource getRefAPISource() {
+        Map<String, Reference> referenceMap = buildReferenceNameToReferenceTable(pipelineOptions, referenceSetID);
+        return buildReferenceNameToIdTableFromMap(referenceMap);
+    }
 
-        catch ( IOException e ) {
-            throw new UserException("Error while looking up references for reference set " + referenceSet.getId(), e);
-        }
+    /**
+     * Given a reference set ID, get a SAMSequenceDictionary that lists its contigs and their length.
+     * Note that this does NOT use the referenceContig order from the API. Instead, we'll sort them
+     * and try to match the read sequence dictionary order, if given.
+     *
+     * @param pipelineOptions - are used to get the credentials necessary to call the Genomics API
+     * @param referenceSetID - the ID of the reference set to use
+     * @param optReadSequenceDictionaryToMatch - (optional) the sequence dictionary of the reads, we'll match its order if possible.
+     * @return a SAMSequenceDictionary that lists the referenceset's contigs and their length.
+     */
+    public SAMSequenceDictionary getReferenceSequenceDictionary(final PipelineOptions pipelineOptions, String referenceSetID, final SAMSequenceDictionary optReadSequenceDictionaryToMatch) {
+        Map<String, Reference> referenceMap = buildReferenceNameToReferenceTable(pipelineOptions, referenceSetID);
+        return getReferenceSequenceDictionaryFromMap(referenceMap, optReadSequenceDictionaryToMatch);
+    }
 
-        return referenceNameToIdTable;
+    /**
+     * Given a reference set ID, get a SAMSequenceDictionary that lists its contigs and their length.
+     * Note that this does NOT use the referenceContig order from the API. Instead, we'll sort them
+     * and try to match the read sequence dictionary order, if given.
+     *
+     * Also produces a table from reference name (sadly not standard names like GRCh37),
+     * to IDs need for Google Genomics API calls. The table is produced via a query to get the mapping. The query
+     * currently takes ~10 seconds, so this table should be cached and no produced per query to get the reference bases.
+     * This is clumsy and should be refactored with better Genomics API use (issue 643).
+     * @param pipelineOptions options needed for credentials to produce a Genomics class for the API call.
+     * @param referenceSetID the ID of the reference set, e.g., EOSt9JOVhp3jkwE.
+     * @param optReadSequenceDictionaryToMatch - (optional) the sequence dictionary of the reads, we'll match its order if possible.
+     * @return returns KV(a SAMSequenceDictionary that lists the referenceset's contigs and their length, a mapping from reference name to String ID)
+     */
+    public KV<SAMSequenceDictionary, Map<String, String>> getReferenceSequenceDictionaryAndBuildReferenceNameToIdTable(final PipelineOptions pipelineOptions, String referenceSetID, SAMSequenceDictionary optReadSequenceDictionaryToMatch) {
```

Comment:
I have major performance objections. However, I have since (a) fixed the underlying performance problem and (b) refactored this code so it still does only one API round-trip, but it does so in a way that should be easier to read.